File: core/src/commonMain/kotlin/eldoria/core/model/Item.kt
package eldoria.core.model

import kotlinx.serialization.Serializable

/**
 * WEAPON/ARMOR(chest)/OFFHAND/HEAD/RING/AMULET are the six equip slots on
 * PlayerCharacter (see equippedWeapon/equippedOffhand/equippedArmor/
 * equippedHead/equippedRing/equippedAmulet) -- ARMOR specifically means the
 * chest slot, kept its original name to avoid churning every existing call
 * site for a cosmetic rename. RING/AMULET carry their bonus via
 * `magicEffect` rather than `armorClassBonus`, reusing the same mechanism
 * legendary items already use instead of adding parallel bonus fields.
 */
enum class ItemKind { WEAPON, ARMOR, OFFHAND, HEAD, RING, AMULET, QUEST_ITEM, TRINKET, MATERIAL, BOAT }

/**
 * Any physical item: weapon, armor, quest item, or misc trinket. Combat and
 * shop stats are all dice-derived (see StatGenerator) -- value and durability
 * included, so "wear and tear" is a real mechanical value, not flavor text.
 */
@Serializable
data class Item(
    val name: String,
    val kind: ItemKind,
    val tier: Int,
    val damage: DiceFormula? = null,
    val armorClassBonus: Int? = null,
    val magicEffect: MagicEffect? = null,
    val value: Int,
    val maxDurability: Int,
    val currentDurability: Int = maxDurability,
    val isLegendary: Boolean = false,
    /** BOAT-only: fitted cannons add a bonus damage volley each round of ship combat, see Game.kt's shipEncounter. */
    val hasCannons: Boolean = false,
    /** WEAPON-only: a chance to inflict this on a successful hit -- see Game.kt's `attack` and StatGenerator's legendary-weapon roll. */
    val inflictsStatus: StatusEffect? = null,
) {
    val isBroken: Boolean get() = currentDurability <= 0
    fun worn(amount: Int): Item = copy(currentDurability = (currentDurability - amount).coerceAtLeast(0))
    fun repaired(): Item = copy(currentDurability = maxDurability)
}

File: eldoria/models/item.py
"""Any physical item: weapon, armor, quest item, or misc trinket."""
from __future__ import annotations

from dataclasses import dataclass, replace
from enum import Enum

from eldoria.models.dice import DiceFormula
from eldoria.models.stat_block import MagicEffect
from eldoria.models.status_effect import StatusEffect


class ItemKind(Enum):
    """WEAPON/ARMOR(chest)/OFFHAND/HEAD/RING/AMULET are the six equip slots on PlayerCharacter."""

    WEAPON = "WEAPON"
    ARMOR = "ARMOR"
    OFFHAND = "OFFHAND"
    HEAD = "HEAD"
    RING = "RING"
    AMULET = "AMULET"
    QUEST_ITEM = "QUEST_ITEM"
    TRINKET = "TRINKET"
    MATERIAL = "MATERIAL"
    BOAT = "BOAT"
    CONSUMABLE = "CONSUMABLE"


@dataclass(frozen=True)
class Item:
    name: str
    kind: ItemKind
    tier: int
    value: int
    max_durability: int
    damage: DiceFormula | None = None
    armor_class_bonus: int | None = None
    magic_effect: MagicEffect | None = None
    current_durability: int | None = None
    is_legendary: bool = False
    has_cannons: bool = False
    inflicts_status: StatusEffect | None = None
    is_compounding: bool = False
    """WEAPON-only: each consecutive hit on the same target doubles the last hit's damage (capped). See commands.attack."""
    heal_amount: int | None = None
    """CONSUMABLE-only: health restored on use, see commands.use_item."""

    def __post_init__(self):
        if self.current_durability is None:
            object.__setattr__(self, "current_durability", self.max_durability)

    @property
    def is_broken(self) -> bool:
        return self.current_durability <= 0

    def worn(self, amount: int) -> "Item":
        return replace(self, current_durability=max(0, self.current_durability - amount))

    def repaired(self) -> "Item":
        return replace(self, current_durability=self.max_durability)
